CI: Add IPsec key rotation test

jschwinger233 opened the pull request:
This PR adds IPsec key rotation test in `conformance-ipsec-e2e.yaml` (ci-ipsec-e2e). A new github action `conn-disrupt-test` is also used to reduce duplication for this test as well as IPsec upgrade test.

Fixes: #26350

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

bimmlerd commented on .github/workflows/tests-ipsec-upgrade.yaml:
```diff
@@ -342,32 +336,11 @@ jobs:
 
-      - name: Test Cilium after upgrade (${{ matrix.name }})
-        uses: cilium/little-vm-helper@6a3d1797af8fab1a49f1c3d09afe3230aea559b6 # v0.0.11
+      - name: Downgrade Cilium ${{ env.cilium_stable_version }} & Test (${{ matrix.name }})
```
nit:
```suggestion
      - name: Downgrade Cilium to ${{ env.cilium_stable_version }} & Test (${{ matrix.name }})
```

bimmlerd commented on .github/workflows/tests-ipsec-upgrade.yaml:
```diff
@@ -322,16 +321,11 @@ jobs:
-      - name: Upgrade Cilium (${{ matrix.name }})
-        uses: cilium/little-vm-helper@6a3d1797af8fab1a49f1c3d09afe3230aea559b6 # v0.0.11
+      - name: Upgrade Cilium & Test (${{ matrix.name }})
+        uses: ./.github/actions/conn-disrupt-test
```
I'm under the impression this action gets executed from the PR branch. If so, as a malicious actor I can put stuff in the commands of the action and get code execution in a PR, no?

jschwinger233 replied:
I'm not sure but I expect only a [member of cilium](https://github.com/cilium/community/blob/main/ladder/members.yaml) has permission to create a branch?

bimmlerd replied:
That's true, but I think it'll also checkout a branch of a fork, which anyone can create. See https://github.com/cilium/cilium/pull/27203/files#diff-b77d9c4420c99e9bbac5afc52278677e846739ed4b7b4f7c1efdb24f1b69a249L216

pchaigno replied:
I believe David is correct. The only other time we execute a checked out script (`./contrib/scripts/kind.sh`) in this workflow is within a VM (LVH). Might be worth getting a confirmation from @nbusseneau.

If this is insecure, then I guess our (only?) option is to move this new GitHub action to its own repository. Or can we use a specific version (specific SHA) of this new GitHub action without moving it elsewhere?

jschwinger233 replied:
Seems we can references a subdirectory in a public GitHub repository at a specific branch, ref, or SHA.

nbusseneau replied:
That is correct, we must be careful with how local `uses` are used in privileged contexts such as `workflow_dispatch`-based workflows, since they have access to repository secrets and also have extended permissions.

If you grep for `uses: .` you can inspect how do we do it for the existing composite actions (for env variables and EKS configuration): we make sure the local `uses` are only done in the "trusted" part of the workflow, and we do not use them in the "untrusted" part (which is everything after we checked out the PR code, see line 234 in this file for example).

I was thinking the other day that we should remove local `uses` entirely and switch to a static reference anyway, that would help with not having to concern ourselves with this specific issue for maintenance and reviews, so I would advocate for doing so here:

```suggestion
        uses: ${{ github.repository }}/.github/actions/conn-disrupt-test@${{ inputs.context-ref || github.sha }}
```

This should pull the composite actions from the appropriate context based on trigger (if `workflow_dispatch`, `input.context-ref` will be either the PR target branch or the PR branch itself depending on whether the PR comes from a fork or not, and if `schedule`, `github.sha` will be used as fallback).

jschwinger233 replied:
Unfortunately `uses: ${{ github.repository }}/.github/actions/conn-disrupt-test@${{ inputs.context-ref || github.sha }}` is invalid: https://github.com/cilium/cilium/actions/runs/5889885321

```
Invalid workflow file: .github/workflows/tests-ipsec-upgrade.yaml#L333
The workflow is not valid. .github/workflows/tests-ipsec-upgrade.yaml (Line: 333, Col: 15): Unrecognized named-value: 'github'. Located at position 1 within expression: github.repository .github/workflows/tests-ipsec-upgrade.yaml (Line: 352, Col: 15): Unrecognized named-value: 'github'. Located at position 1 within expression: github.repository
```

So I decide to create an independent PR https://github.com/cilium/cilium/pull/27567 to add this new action, then we can refer to the action using `cilium/cilium/.github/actions/conn-disrupt-test@SHA` to address the security concern.

bimmlerd reviewed (commented):
looks good - one security concern I'm not entirely sure about, I'm sure github-sec will chime in :)

pchaigno commented on .github/actions/conn-disrupt-test/action.yaml:
```diff
@@ -0,0 +1,48 @@
+        cmd: |
+          cd /host/
+          # Disable no-missed-tail-calls due to https://github.com/cilium/cilium/issues/26739
+          # Disable check-log-errors due to https://github.com/cilium/cilium-cli/issues/1858
```
Downside of the new GitHub action is that we end up disabling those two checks for key rotations as well :disappointed: Any way to avoid that?

jschwinger233 replied:
Yeah I can avoid these two hardcoded disabling by introducing one more input parameter, so caller can pass flags like
```
uses: conn-disrupt-test
with:
  extra-connectivity-test-flags: "--test '!no-missed-tail-calls'"
```
Action can therefore run connectivity test with caller-specified flags.

For key rotation test, caller just don't specify any extra flags so as to run full tests.

pchaigno replied:
Sounds good to me :+1:

pchaigno commented on .github/workflows/conformance-ipsec-e2e.yaml:
```diff
@@ -278,7 +278,40 @@ jobs:
+
+            # Re-create cilium pods to reload new ipsec key
+            kubectl -n kube-system delete pod -l k8s-app=cilium
+            ./cilium-cli status --wait
```
This is not needed. There is a watcher that will detect the updated key file and reload the key automatically. I believe it runs once per minute or so.

pchaigno commented on .github/workflows/conformance-ipsec-e2e.yaml:
```diff
@@ -278,7 +278,40 @@ jobs:
+              fi
+              echo "Waiting until key rotation completes"
+              sleep 30s
+            done
```
By default the key rotation cleanup delay is 5min, so you could probably do a 4min sleep before starting to actively poll for the number of keys.

pchaigno reviewed (commented):
Apart from the potential security issue we need to figure out, this looks great!

The fact that the test is passing however makes me very suspicious. We know that key rotations currently cause XfrmOutPolBlock errors (which @jrfastab is working on fixing). How can the test be passing? :confused:

jschwinger233 committed:
ci-ipsec-e2e: Add IPsec key rotation test

This commit adds a step in conformance-ipsec-e2e to perform IPsec key
rotation and conn-disrupt-test.

The commands to perform IPsec key rotation are copied from
https://docs.cilium.io/en/latest/security/network/encryption-ipsec/#key-rotation with a few improvements:
* To pass commands with "$" character to conn-disrupt-test action, "$"
  must be escaped as "\$".
* To have a new IPsec key in a bare VM, a string of hex is hard coded
  without using xxd.

Fixes: #26350

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

jschwinger233 committed:
ci-ipsec-upgrade: Use conn-disrupt-test action for tests-ipsec-upgrade

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>
